src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,349 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_GRPCSERVER_H_INCLUDED
+#define RIPPLE_CORE_GRPCSERVER_H_INCLUDED
+
+#include <ripple/rpc/Role.h>
+#include <ripple/rpc/Context.h>
+#include <ripple/rpc/GRPCHandlers.h>
+#include <ripple/rpc/impl/RPCHelpers.h>
+#include <ripple/rpc/impl/Handler.h>
+#include <ripple/rpc/impl/Tuning.h>
+#include <ripple/net/InfoSub.h>
+#include <ripple/protocol/ErrorCodes.h>
+#include <ripple/core/JobQueue.h>
+#include <ripple/resource/Charge.h>
+
+#include <grpcpp/grpcpp.h>
+#include "rpc/v1/xrp_ledger.grpc.pb.h"
+
+namespace ripple {
+
+//Interface that CallData implements
+class Processor
+{
+    public:
+
+    virtual ~Processor() {}
+
+    // process a request that has arrived. Can only be called once per instance
+    virtual void process() = 0;
+
+    //store an iterator to this object
+    //all Processor objects are stored in a std::list, and the iterator points
+    //to that objects position in the list. When object finishes processing a 
+    //request, the iterator is used to delete the object from the list
+    virtual void set_iter(
+            std::list<std::shared_ptr<Processor>>::iterator const& it) = 0;
+
+    //get iterator to this object. see above comment
+    virtual std::list<std::shared_ptr<Processor>>::iterator get_iter() = 0;
```

Comment:
You are right that storing the iterator is just an optimization used when deleting the object from the collection.

I am hesitant to add a search through a collection in the main event loop. `handleRpcs()` is not doing very much work; everything is an `O(1)` call. However, adding a search through a collection here, while not extremely expensive, is asymptotically more expensive than what is currently happening in `handleRpcs()`, and is an `O(n)` call, where `n` is the size of the collection. So, this would increase the runtime complexity of each iteration of the loop from `O(1)` to `O(n)`. Sure, the collection will be probably be small most of the time, but there is no guarantee that the collection will be small. Adding extra work to `handleRpcs()` is going to affect the performance of every rpc call.

I am also hesitant to keep the processors in a dictionary, since erasing from the dictionary will not reclaim the memory that is used by the dictionary. If there is a spike of concurrent requests, the dictionary could end up very large, and then will never decrease in size, even once the requests are processed.

boost::intrusive seems like a good fit here, since we wouldn't need to store the iterator, but could compute it in constant time. However, boost::intrusive doesn't allow us to store `shared_ptr` [https://www.boost.org/doc/libs/1_71_0/doc/html/intrusive/using_smart_pointers.html](https://www.boost.org/doc/libs/1_71_0/doc/html/intrusive/using_smart_pointers.html). I'm not sure if the lifetime of these objects could be managed in a clean way without using shared pointers.

Just curious, with the dictionary, what would you use as a key?

Reply:
@cjcobb23 I was thinking of using the address of the object as the key. But thinking about it some more, I'd just use a `std::set<shared_ptr<..>` or or `std::unordered_set<shared_ptr<...>>`; This is really a set, not a dictionary.

`std::{map, set}` is tree based, so it should reclaim the memory (but allocates memory on insertion). `std::{unordered_map, unordered_map}` is a hash table and I'm not sure what the behavior is for reclaiming memory in those containers. You might be right that the memory would go unclaimed - I believe Microsoft at least uses a `vector` as the underlying container for buckets, so that would go unclaimed, as you sat. I still think the number of in-flight requests is going to be so small that it will never matter - both for a linear search and for extra memory. 

I'd go with an `unordered_set<shared_ptr<...>>`. If you're really worried about reclaiming memory, you can put a check in to see if it's ever wasting more than than X megs, and if it is reclaim the memory (probably my moving the pointers into another container and swapping with that container). But I'm still a little skeptical that it's a real problem.

Reply:
@seelabs After thinking about this, I would prefer going with a vector:
1. The cache friendliness will make searching and erasing very fast, especially when the number of in-flight requests is small. 
2. A vector is going to be the most space efficient of all of these data structures; it uses less memory overall, and results in less fragmentation.

I think if we consider the most typical situation, with not many concurrent requests, a vector is actually going to give the best performance and memory footprint, better than what I have currently probably (because of data locality).

About reclaiming the memory, I am not sure how necessary it is. My original worry was that someone could try to DDOS the server, causing this vector to greatly expand in size, after which the memory will never be released. However, the vector would be of type `vector<shared_ptr<...>>`, and `shared_ptr<...>` is not a large object; I think usually its just the sizeof two pointers, so 16 bytes on 64 bit system. So for this vector to even grow to a MB, we would need > 60k requests concurrently (MB/16 bytes), which I'm not sure is even possible. The requests complete in a few millseconds, and a MB is not even that big anyways. Do you think its even necessary to check the memory usage and reclaim if necessary?

Reply:
@cjcobb23 No, I don't think it's even necessary to check the memory usage and reclaim. I think a `std::vector<shared_ptr<...>>` is a good solution.

Reply:
@cjcobb23 While I'm OK with `std::vector` I'd probably use `unordered_set` for this variable. I'm not worried about performance. I'd like to use a set because it shows the constrains on how this variable is used better than a vector does.

Reply:
@seelabs if we are really not concerned about performance or memory usage, then I agree that `unordered_set` is the most natural data structure to use here. You've eased my worries about performance, so I'll just use `unordered_set`

Reply:
@seelabs using `unordered_set` here may be a bit clunky though. The elements of the set are `std::shared_ptr<Processor>`, but `cq_.Next()` only gives us raw pointers. I would have to use `shared_from_this()` to create an appropriate `shared_ptr` as a temporary, and then pass that shared_ptr to `unordered_set.erase()`.

I could use a `std::unordered_map<Processor*,std::shared_ptr<Processor>>`, where the keys are the actual address of the object (which `cq_.Next()` gives us), but I think that is a bit clunky as well and I would prefer a `std::vector<std::shared_ptr<Processor>>` here.

Reply:
@seelabs changed this to a vector, and added a comment